tests/e2e/tests/pilot/mcp_test.go

```diff
@@ -0,0 +1,137 @@
+// Copyright 2018 Istio Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package pilot
+
+import (
+	"fmt"
+	"io/ioutil"
+	"net"
+	"os"
+	"os/exec"
+	"testing"
+
+	"code.cloudfoundry.org/copilot/testhelpers"
+	"github.com/onsi/gomega"
+	"github.com/onsi/gomega/gbytes"
+	"github.com/onsi/gomega/gexec"
+
+	"istio.io/istio/mixer/test/client/env"
+	"istio.io/istio/pilot/pkg/model"
+	mockmcp "istio.io/istio/tests/e2e/tests/pilot/mock/mcp"
+	"istio.io/istio/tests/util"
+)
+
+const (
+	pilotDebugPort = 5555
+	pilotGrpcPort  = 15010
+	mcpServerAddr  = "127.0.0.1:15014"
+)
+
+func TestPilotMCPClient(t *testing.T) {
+	g := gomega.NewGomegaWithT(t)
+
+	t.Log("building & starting mock mcp server...")
+	mcpServer, err := runMcpServer(g, t)
+	g.Expect(err).NotTo(gomega.HaveOccurred())
+
+	t.Log("building pilot...")
+	istioConfigDir := testhelpers.TempDir()
+	pilotSession, err := runPilot(istioConfigDir, mcpServerAddr, pilotGrpcPort, pilotDebugPort)
+	g.Expect(err).NotTo(gomega.HaveOccurred())
+
+	t.Log("checking if pilot ready")
+	g.Eventually(pilotSession.Out, "10s").Should(gbytes.Say(`READY`))
+	fmt.Println(string(pilotSession.Out.Contents()))
+
+	t.Log("run edge router envoy...")
+	gateway := runEnvoy(t, pilotGrpcPort, pilotDebugPort)
+
+	defer func() {
+		mcpServer.Close()
+		pilotSession.Kill()
+		gateway.TearDown()
+	}()
+
+	t.Log("check that envoy is listening on the configured gateway...")
+	gatewayResource := fmt.Sprintf("127.0.0.1:%s", "8099")
+	g.Eventually(func() error {
+		_, err := net.Dial("tcp", gatewayResource)
+		return err
+	}, "180s", "1s").Should(gomega.Succeed())
+}
+
+func runMcpServer(g *gomega.GomegaWithT, t *testing.T) (*mockmcp.Server, error) {
+	supportedTypes := []string{
+		fmt.Sprintf("type.googleapis.com/%s", model.VirtualService.MessageName),
+		fmt.Sprintf("type.googleapis.com/%s", model.Gateway.MessageName),
+		fmt.Sprintf("type.googleapis.com/%s", model.ServiceEntry.MessageName),
+		fmt.Sprintf("type.googleapis.com/%s", model.EnvoyFilter.MessageName),
+	}
+
+	server, err := mockmcp.NewServer(mcpServerAddr, supportedTypes)
+	if err != nil {
+		return nil, err
+	}
+
+	return server, nil
+}
+
+func runPilot(istioConfigDir, mcpServerAddr string, grpcPort, debugPort int) (*gexec.Session, error) {
+	path, err := gexec.Build("istio.io/istio/pilot/cmd/pilot-discovery")
+	if err != nil {
+		return nil, err
+	}
```

Comment:
Can you extend/copy the approach taken by [initLocalPilotTestEnv](https://github.com/istio/istio/blob/d58b31cca8ea5ab05dd97bb7a2403ea83c6b5af9/pilot/pkg/proxy/envoy/v2/xds_test.go#L105) from the `istio.io/istio/pilot/pkg/proxy/envoy/v2` package?